File: src/main/java/club/sk1er/mods/levelhead/renderer/LevelheadTag.java
package club.sk1er.mods.levelhead.renderer;

import gg.essential.api.utils.JsonHolder;

import java.util.UUID;

public class LevelheadTag {
    private LevelheadComponent header;
    private LevelheadComponent footer;
    private final UUID owner;

    public LevelheadTag(UUID owner) {
        this.owner = owner;
    }

    public LevelheadComponent getHeader() {
        return header;
    }

    public LevelheadComponent getFooter() {
        return footer;
    }

    public void construct(JsonHolder holder) {
        if (header == null) this.header = build(holder, true);
        if (footer == null) this.footer = build(holder, false);
    }

    private LevelheadComponent build(JsonHolder holder, boolean isHeader) {
        String seek = isHeader ? "header" : "footer";
        JsonHolder json = holder.optJSONObject(seek);

        LevelheadComponent component = new LevelheadComponent(json.defaultOptString(seek, "UMM BIG ERROR REPORT TO SK1ER"));
        boolean custom = json.optBoolean("custom");

        component.setCustom(custom);
        if (custom && isHeader && !holder.optBoolean("exclude")) {
            component.setValue(component.getValue() + ": ");
        }

        if (json.optBoolean("chroma")) {
            component.setChroma(true);
        } else if (json.optBoolean("rgb")) {
            component.setRgb(true);
            component.setAlpha(json.optInt("alpha"));
            component.setRed(json.optInt("red"));
            component.setBlue(json.optInt("blue"));
            component.setGreen(json.optInt("green"));
        } else {
            component.setColor(json.optString("color"));
        }

        return component;
    }

    public String getString() {
        return header.getValue() + footer.getValue();
    }

    @Override
    public String toString() {
        return "LevelheadTag{" +
            "header=" + header +
            ", footer=" + footer +
            ", owner=" + owner +
            '}';
    }
}

File: src/main/kotlin/club/sk1er/mods/levelhead/display/LevelheadTag.kt
package club.sk1er.mods.levelhead.display

import club.sk1er.mods.levelhead.Levelhead
import com.google.gson.JsonObject
import java.awt.Color
import java.util.*

class LevelheadTag(val owner: UUID) {
    var header: LevelheadComponent = LevelheadComponent()
    var footer: LevelheadComponent = LevelheadComponent()

    fun getString() = "${header.value}${footer.value}"

    override fun toString(): String = "LevelheadTag{header=$header, footer=$footer, owner=$owner}"

    fun clone(): LevelheadTag = Levelhead.gson.fromJson(Levelhead.gson.toJson(this), LevelheadTag::class.java)

    companion object {
        fun build(owner: UUID, block: LevelheadTagBuilder.() -> Unit) =
            LevelheadTagBuilder(owner).apply(block).tag
    }

    class LevelheadTagBuilder(owner: UUID) {
        val tag = LevelheadTag(owner)

        fun header(block: LevelheadComponent.() -> Unit) =
            tag.header.apply(block)

        fun footer(block: LevelheadComponent.() -> Unit) =
            tag.footer.apply(block)
    }

    class LevelheadComponent {
        var value: String = ""
            set(value) { field = value.replace("&", "\u00a7") }
        var color: Color = Color.WHITE
        var chroma: Boolean = false

        override fun toString(): String = "LevelheadComponent{value='$value', color='${color}', chroma=$chroma}"
    }
}